OSX - support POWER events suspend(ed)/resume in wxWdgets

tisj opened the pull request:
Register observers with notifications center, with callbacks for NSWorkspaceWillSleepNotification and NSWorkspaceDidWakeNotification notifications.

These callbacks inform the top level window to handle POWER SUSPENDED and/or RESUME.

tisj committed:
Add wxCocoaPowerModule and CocoaPowerEventsObserver

Support POWER events suspend(ed)/resume in wxWdgets.

Register observers with notifications center, with callbacks for
NSWorkspaceWillSleepNotification and NSWorkspaceDidWakeNotification
notifications.

These callbacks inform the top level window to handle POWER SUSPENDED
and/or RESUME.

Signed-off-by: Tijs Van Buggenhout <tijs.van.buggenhout@axsguard.com>

vadz commented on src/osx/cocoa/power.mm:
```diff
@@ -105,3 +110,81 @@ bool UpdatePowerResourceUsage(wxPowerResourceKind kind, const wxString& reason)
+#import <Cocoa/Cocoa.h>
+#import "wx/power.h" // Assuming you have wxPowerEvent defined or a custom event
+
+@interface CocoaPowerEventsObserver : NSObject
```
This probably should be called `wxCocoaPowerEventsObserver` for consistency and also maybe to avoid any conflicts with application-defined identifier when using static linking (I'm not sure what are the visibility rules for Objective C interfaces, but using "wx" prefix can never hurt).

vadz commented on src/osx/cocoa/power.mm:
```diff
@@ -105,3 +110,81 @@ bool UpdatePowerResourceUsage(wxPowerResourceKind kind, const wxString& reason)
+
+    void OnExit() override
+    {
+         if (m_powerObserver != nullptr) {
```
Minor style pick, wx style puts braces on their own lines:

```suggestion
         if (m_powerObserver != nullptr)
         {
```

vadz reviewed (commented):
Nice, thanks for implementing this!

The documentation in `interface/wx/power.h` should be updated to mention that this functionality is now available under macOS too.

I also wonder if we could/should delay `CocoaPowerEventsObserver` construction until `wxPowerResourceBlocker` is instantiated: this is required to get these events under Linux and even if it isn't under macOS, it could be nice to encourage writing more portable code. What do you think?

tisj commented:
If i understand it correctly the wxResourcePowerBlocker is defined in include/wx/power.h and gets instantiated when the application includes this file (directly or indirectly)?

In Linux these events are observed on the DBus system bus via login manager, but that implementation resides in src/gtk/power.cpp - and thus part of wx library.

No sure how the magic is happening here ;)

vadz commented:
> If i understand it correctly the wxResourcePowerBlocker is defined in include/wx/power.h and gets instantiated when the application includes this file (directly or indirectly)?

No, it's just a normal class, there is no magic involved. The application needs to create an object of this class in order to receive power events under Linux/GTK, see the example in `wxPowerEvent` documentation.

My thinking is that we could also only send these events in macOS implementation if/after `wxPowerResourceBlocker` was/is created, even if this is unnecessary under this platform. But by forcing the application to do it we would

1. Avoid creating the observer unnecessarily (for the applications that don't care about such events).
2. Ensure that the code using these events developed under macOS also works under Linux (instead of silently doing nothing there).

P.S. Please don't force push until and unless you're pushing the very final version of the PR, force pushing makes it difficult to see what has changed, so please use "fixup" commits as created by `git commit --fixup` or `--squash` and then force push just once at the end. TIA!

tisj committed:
fixup! Add wxCocoaPowerModule and CocoaPowerEventsObserver

https://github.com/wxWidgets/wxWidgets/pull/25778#discussion_r2334034960

Signed-off-by: Tijs Van Buggenhout <tijs.van.buggenhout@axsguard.com>

tisj committed:
fixup! Add wxCocoaPowerModule and CocoaPowerEventsObserver

Enable Power Events on request.

Create a unique wxCocoaPowerEventsObserver when system power resources
are acquired.

Wrap the observer in a C++ class for handling via std::unique_ptr
(sizeof) and automatic init/release of resources.

Signed-off-by: Tijs Van Buggenhout <tijs.van.buggenhout@axsguard.com>

vadz reviewed (commented):
Great, thanks for implementing this!

AFAICS this should work exactly as desired, except perhaps that we should destroy the power observer when the counter drops to 0 too.

But we obviously need to fix the build first and maybe address the remarks below.

Thanks!

tisj commented on src/osx/cocoa/power.mm:
```diff
@@ -13,18 +13,125 @@
 #include <IOKit/pwr_mgt/IOPMLib.h>
 
+#import <Cocoa/Cocoa.h>
+#import "wx/power.h" // Assuming you have wxPowerEvent defined or a custom event
```
It was generated by gemini

<img width="1275" height="576" alt="image" src="https://github.com/user-attachments/assets/6c9b205a-b3a4-4d37-a586-a65c97b8a593" />

Probably picked it up from some other project I guess

tisj replied:
I removed the line

tisj commented on src/osx/cocoa/power.mm:
```diff
@@ -13,18 +13,125 @@
+{
+public:
+    wxCocoaPowerEventsObserverWrapper()
+        : m_powerEventsObserver( [[CocoaPowerEventsObserver alloc] init] )
```
:ashamed: .. I used an older implementation (before the rename) to compile and test, then cp to git and thought to have done the rename completely

tisj committed:
fixup! Add wxCocoaPowerModule and CocoaPowerEventsObserver

Replace std::unique_ptr with wxCFRef

Signed-off-by: Tijs Van Buggenhout <tijs.van.buggenhout@axsguard.com>

vadz reviewed (approved):
Looks good now, thanks!

Will merge soon.

tisj commented:
Just a question... When you would acquire the system power resource on windows (delay),it would just return true without incrementing g_powerResourceSystemRefCount.

If then you release that resource, it will see g_powerResourceSystemRefCount is 0 and issue a fail "System power resource was not acquired", which is not desirable when writing platform independent code.

Wouldn't it be better to always increment on acquire and decrement on release?

vadz commented:
> Just a question... When you would acquire the system power resource on windows (delay),it would just return true without incrementing g_powerResourceSystemRefCount.
> 
> If then you release that resource, it will see g_powerResourceSystemRefCount is 0 and issue a fail "System power resource was not acquired", which is not desirable when writing platform independent code.
> 
> Wouldn't it be better to always increment on acquire and decrement on release?

Yes, indeed, there is a problem here, thanks for raising it. But handling `wxPOWER_DELAY` as `wxPOWER_PREVENT` under MSW doesn't look great either... The simplest non-horrible fix seems to be to pass `blockKind` to `Release()` too.

But maybe this just shows that having `wxPowerBlockKind` was a bad idea in the first place as it's not really orthogonal to `wxPowerResourceKind` under any platform: under Linux and macOS (with your patch) `wxPOWER_DELAY` can't be combined with `wxPOWER_RESOURCE_SCREEN`, so maybe we should just drop `wxPowerBlockKind` completely and add `wxPOWER_RESOURCE_SYSTEM_DELAY` or something like this with a better name which would work as `wxPOWER_DELAY` does now?

Any change like this would/should be done in a separate PR, of course.
